selfhost/typechecker.jakt

```diff
@@ -12681,3 +12661,159 @@ fn dump_scope(anon scope_id: ScopeId, anon program: &CheckedProgram, indent: i64
         dump_scope(id, program, indent: cindent + 2)
     }
 }
+
+enum BindingKey {
+    Found(usize) // found equivalent pattern at an index
+    New([String:VarId]) // need to make a new case.
+}
+
+fn search_empty_pattern(cases: ArraySlice<CheckedMatchCase>) -> BindingKey {
+    for idx in 0..cases.size() {
+        if cases[idx].bindings.is_empty() {
+            return BindingKey::Found(idx)
+        }
+    }
+    return BindingKey::New([:])
+}
+
+// The state that is kept while discovering bindings from a pattern inside a
+// match, which discards possible aliasing cases on each binding discovery. It
+// is used by the typechecker to detect when the body of a case must be
+// typechecked with new types for a certain pattern. The `cases` slice is the
+// subset of cases that the pattern can refer to, which are the checked cases
+// from previous patterns that share the same body. This invariant is easy to
+// miss, but `MatchBuilder` (below) helps in enforcing it.
+// NOTE: The `correct` set may be implemented with a bit array with size cases.size()
+enum BindingKeyBuilder {
+    Empty(cases: ArraySlice<CheckedMatchCase>)
+    Known(built: [String:VarId], correct: {usize}, cases: ArraySlice<CheckedMatchCase>)
+    New(built: [String:VarId])
+
+    fn finish(this) -> BindingKey => match this {
+        Empty(cases) => search_empty_pattern(cases)
+        Known(built, correct, cases) => {
+            // From the remaining possibilities, find one that is exactly the
+            // size of the binding list.
+            for idx in correct {
+                if cases[idx].bindings.size() == built.size() {
+                    return BindingKey::Found(idx)
+                }
+            }
+            yield BindingKey::New(built)
+        }
+        New(built) => BindingKey::New(built)
+    }
+
+    fn submit(
+        mut this
+        anon name: String
+        anon var_id: VarId
+        program: CheckedProgram
+    ) -> BindingKeyBuilder => match this {
+        Empty(cases) => {
+            let correct = build_correct_set(0..cases.size(), name, var_id, cases, program)
+            yield BindingKeyBuilder::from_set(correct, built: [name:var_id], cases)
+        }
+        Known(built: built_, correct, cases) => {
+            mut built = built_
+            built[name] = var_id
+
+            let next_correct = build_correct_set(correct.iterator(), name, var_id, cases, program)
+
+            yield BindingKeyBuilder::from_set(correct: next_correct, built, cases)
+        }
+        New(built: built_) => {
+            mut built = built_
+            built[name] = var_id
+            yield BindingKeyBuilder::New(built)
+        }
+    }
+
+    // NOTE: This could require IntoIterator, but Range (0..cases.size() in Empty case above) is Iterable, not IntoIterator.
```

Comment:
🤔 aside: maybe we should make `Iterable<T>` implement `IntoIterator<T>`?

Reply:
Just saw #1603. If you end up finding a solution you're comfortable with I'll wait for your fix and then fix this immediately, otherwise I can leave it as a FIXME.

Reply:
Nah, that's a big mess I don't want to get into right now